File: solutions/java/0336.java
class Solution {
  public List<List<Integer>> palindromePairs(String[] words) {
    List<List<Integer>> ans = new ArrayList<>();
    Map<String, Integer> map = new HashMap<>();

    for (int i = 0; i < words.length; ++i) {
      StringBuilder sb = new StringBuilder(words[i]);
      map.put(sb.reverse().toString(), i);
    }

    for (int i = 0; i < words.length; ++i) {
      if (map.containsKey("") && map.get("") != i && isPalindrome(words[i]))
        ans.add(Arrays.asList(i, map.get("")));

      for (int j = 1; j <= words[i].length(); ++j) {
        String l = words[i].substring(0, j);
        String r = words[i].substring(j, words[i].length());
        if (map.containsKey(l) && map.get(l) != i && isPalindrome(r))
          ans.add(Arrays.asList(i, map.get(l)));
        if (map.containsKey(r) && map.get(r) != i && isPalindrome(l))
          ans.add(Arrays.asList(map.get(r), i));
      }
    }

    return ans;
  }

  private boolean isPalindrome(String s) {
    int l = 0;
    int r = s.length() - 1;

    while (l < r)
      if (s.charAt(l++) != s.charAt(r--))
        return false;

    return true;
  }
}

File: solutions/cpp/0336.cpp
class Solution {
 public:
  vector<vector<int>> palindromePairs(vector<string>& words) {
    vector<vector<int>> ans;
    unordered_map<string, int> map;

    for (int i = 0; i < words.size(); ++i) {
      string word = words[i];
      reverse(begin(word), end(word));
      map[word] = i;
    }

    for (int i = 0; i < words.size(); ++i) {
      if (map.count("") && map[""] != i && isPalindrome(words[i]))
        ans.push_back({i, map[""]});

      for (int j = 1; j <= words[i].length(); ++j) {
        const string l = words[i].substr(0, j);
        const string r = words[i].substr(j, words[i].length() - j);
        if (map.count(l) && map[l] != i && isPalindrome(r))
          ans.push_back({i, map[l]});
        if (map.count(r) && map[r] != i && isPalindrome(l))
          ans.push_back({map[r], i});
      }
    }

    return ans;
  }

 private:
  bool isPalindrome(const string& s) {
    int l = 0;
    int r = s.length() - 1;
    while (l < r)
      if (s[l++] != s[r--])
        return false;
    return true;
  }
};